frontend/app/routes/protected/in-person/state-machine-service.ts

```diff
@@ -0,0 +1,114 @@
+
+import type { Actor } from 'xstate';
+import { createActor } from 'xstate';
+
+import { LogFactory } from '~/.server/logging';
+import { AppError } from '~/errors/app-error';
+import { ErrorCodes } from '~/errors/error-codes';
+import { i18nRoutes } from '~/i18n-routes';
+import type { StateName } from '~/routes/protected/in-person/state-machine';
+import { machine } from '~/routes/protected/in-person/state-machine';
+import { getLanguage } from '~/utils/i18n-utils';
+import { getRouteByFile } from '~/utils/route-utils';
+
+type Machine = typeof machine;
+
+const log = LogFactory.getLogger(import.meta.url);
+
+/**
+ * Creates a new in-person application state machine actor and stores it in the session.
+ *
+ * This function retrieves the `tabId` from the request URL, creates a new XState actor
+ * for the in-person application flow, subscribes to its state changes to persist
+ * the snapshot in the session, and starts the actor.
+ *
+ * @throws {Response} 400 response if the `tabId` is missing in the request.
+ */
+export function createMachineActor(session: AppSession, request: Request): Actor<Machine> {
+  const flow = (session.inPersonFlow ??= {}); // ensure session container exists
+
+  const tabId = new URL(request.url).searchParams.get('tid');
+
+  if (!tabId) {
+    log.warn('Could not find tabId in request; returning 400 response.');
+    throw Response.json('The tabId could not be found in the request', { status: 400 });
+  }
+
+  const actor = createActor(machine);
+  actor.subscribe((snapshot) => void (flow[tabId] = snapshot));
+  log.debug('Created new in-person state machine for session [%s] and tabId [%s]', session.id, tabId);
+
+  return actor.start();
+}
+
+/**
+ * Loads an existing in-person application state machine actor from the session.
+ *
+ * This function retrieves the `tabId` from the request URL, attempts to load a
+ * persisted XState actor snapshot from the session, and recreates the actor
+ * with the loaded snapshot.
+ *
+ * @throws {AppError} If the machine snapshot is not found in the session.
+ * @throws {Response} 400 response if the `tabId` is missing in the request.
+ */
+export function loadMachineActor(session: AppSession, request: Request, state: StateName): Actor<Machine> {
+  const flow = (session.inPersonFlow ??= {}); // ensure session container exists
+
+  const tabId = new URL(request.url).searchParams.get('tid');
+
+  if (!tabId) {
+    // XXX :: GjB :: how should we handle this case?
+    //               Showing an error page seems like bad UX.
+    log.warn('Could not find tabId in request; returning 400 response.');
+    throw Response.json('The tabId could not be found in the request', { status: 400 });
+  }
+
+  if (!flow[tabId]) {
+    // XXX :: GjB :: how should we handle this case?
+    //               Showing an error page seems like bad UX.
+    throw new AppError(
+      'The machine snapshot was not found in session', //
+      ErrorCodes.MISSING_SNAPSHOT,
+    );
+  }
```

Comment:
You log warnings and you do one of the following options:

1. `throw new Response("Not Found", { status: 404 });`
2. `throw i18nRedirect('routes/protected/index.tsx', request);`

Reply:
Re: throwing a redirect

I think this is probably the correct approach, but I think we should consider adding some sort of message/notification so the user isn't surprised when it happens.

Imagine the scenario where a user bookmarks a step and then tries to come back to it later on. If we redirect back to the start, I feel like a message would be useful to at least let the user know why the page she bookmarked isn't being rendered.

Reply:
What do you think about what I did in 07517daeb66a9aa407e7942dde2e3794a2bcb9ac?

Reply:
> Imagine the scenario where a user bookmarks a step and then tries to come back to it later on. If we redirect back to the start, I feel like a message would be useful to at least let the user know why the page she bookmarked isn't being rendered.

That scenario doesn't make sense in an administrative application where things are constantly changing. Additionally, the specific case we're currently working on involves creating a new case, and all routes remain accessible throughout the creation process.

Reply:
I don't disagree, but when the application does something unexpected (like redirect to `index.tsx` seemingly out of the blue), I think it's at least considerate to notify the user why.

My example was contrived and not really realistic, but I still think we should handle the situation gracefully.

Reply:
We do by redirecting.

Reply:
Let me rephrase:

> ...but I still think we should handle the situation gracefully *from the user's perspective*.